qiskit_nature/properties/second_quantization/electronic/integrals/electronic_integrals_utils/two_body_symmetry_conversion.py

```diff
@@ -0,0 +1,176 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Utility functions to detect and transform the index-ordering convention of two-body integrals"""
+
+from enum import Enum
+import numpy
+
+from qiskit_nature import QiskitNatureError
+
+
+def to_chem(two_body_tensor):
+    """
+    Convert the rank-four tensor `two_body_tensor` representing two-body integrals from physicists'
+    , or intermediate, index order to chemists' index order: i,j,k,l -> i,l,j,k
+    """
+    index_order = find_index_order(two_body_tensor)
+    if index_order == IndexType.CHEM:
+        return two_body_tensor
+    if index_order == IndexType.PHYS:
+        chem_tensor = _phys_to_chem(two_body_tensor)
+        return chem_tensor
+    if index_order == IndexType.INT:
+        chem_tensor = _chem_to_phys(two_body_tensor)
+        return chem_tensor
+    else:
+        raise QiskitNatureError(
+            """
+            Unknown index order type, input tensor must be chemists', physicists',
+            or intermediate index order
+            """
+        )
+
+
+def to_phys(two_body_tensor):
+    """
+    Convert the rank-four tensor `two_body_tensor` representing two-body integrals from chemists'
+    , or intermediate, index order to physicists' index order: i,j,k,l -> i,l,j,k
+    """
+    index_order = find_index_order(two_body_tensor)
+    if index_order == IndexType.PHYS:
+        return two_body_tensor
+    if index_order == IndexType.CHEM:
+        phys_tensor = _chem_to_phys(two_body_tensor)
+        return phys_tensor
+    if index_order == IndexType.INT:
+        phys_tensor = _phys_to_chem(two_body_tensor)
+        return phys_tensor
+    else:
+        raise QiskitNatureError(
+            """
+            Unknown index order type, input tensor must be chemists', physicists',
+            or intermediate index order
+            """
+        )
+
+
+def _phys_to_chem(two_body_tensor):
+    """
+    Convert the rank-four tensor `two_body_tensor` representing two-body integrals from physicists'
+    index order to chemists' index order: i,j,k,l -> i,l,j,k
+
+    See `_chem_to_phys`, `_check_two_body_symmetries`.
+    """
+    permuted_tensor = numpy.einsum("ijkl->iljk", two_body_tensor)
+    return permuted_tensor
```

Comment:
This doesn't seem quite right. You cannot simply permute the two-body tensor; reordering terms causes some one-body terms to pop out because of the anticommutation relations. These one-body terms need to be kept track of somehow, otherwise the resulting operator is not equivalent to the original.

Reply:
The way of doing this transformation is dating back to the early days of Qiskit Aqua: https://github.com/Qiskit/qiskit-aqua/commit/d77b4302cd781b88a010f54e0c17bd877b31a6a7

Reply:
All I'm saying is that this is how this conversion has always been done in Nature. Maybe it was just wrong to call it a chemist<->physicist conversion. This PR was started on the assumption that this conversion is correct, but what you're saying would resolve a lot of confusion I used to have around this topic.

This PR is not going to be merged as is anyways, since it will only be integrated in the upcoming refactoring. I converted it to draft until we have more time to look into this.

TL;DR in Aqua and Nature conversion which were called chemist<->physicist conversions always _ignore_ all one-body corrections that would result from them

Reply:
I see, thanks for clarifying. I think PolynomialTensor will make this nice because you could store the correction in the same PolynomialTensor object.

Reply:
Ok, that makes sense. Thanks!

Reply:
These functions implement a relabeling of coefficients. The coefficients have four indices. Each index points to a different Factor in the product of operators in a single term. It leaves each product of operators untouched. I might have one routine that performs computations assuming one index ordering. And another routine that does exactly the same  calculation but expects a different index ordering. These relabeling functions just allow the second routine to do exactly the same thing the first routine did. I think this agrees with what Max said above.